### DISCUSSION TITLE
How to marry Python and Node Virtual Environments, for using ganache (cli) in venv with nodeenv

### DISCUSSION THREAD

**Discussion by n4n0b1t3** [smartcontractkit/full-blockchain-solidity-course-py#938] [Ideas]
For my Python projects I prefer to use a virtual environment. This helps to avoid version conflicts of installed packages and you can then also package the project nicely.
After setting up a Virtual Environment (venv) in a working directory, Virtual Studio Code will ask if you want to activate it for the area. Here you should click on "Yes", so that when you open the terminal the virtual environment is always activated automatically. This can be recognized by the fact that the name of the virtual environment appears in front of the prompt.
But if you want to install a Node JS module like Ganache in this environment, then Ganach will be installed, but the virtual environment doesn't know about it and `ganache --version` runs into an error.

For such cases there is the module [nodevenv](https://nbdime.readthedocs.io/en/latest/nodevenv.html). In the following I describe the (not quite trivial) setup procedure for Windows. A complete example including the setup of python venv is at the end of this article.

Open Powershell **as administrator** and change to the directory of your project.

Activate the Virtual Python Environment (mine is located in the project directory under `.venv`.
`>.venv\Scripts\Activate.ps1`

Install the module nodeenv with
`(.venv) >pip install nodeenv`

Connect nodeenv with your virtual Python environment (might cause your antivirus to start scanning)
`(.venv) >nodeenv -p`

Check if the virtual environment is used for npm with
`(.venv) >npm root -g`
`C:\Users\[...]\.venv\Scripts\node_modules`

Check which modules are already installed (-g global here refers to the virtual environment)
```
(.venv) >npm list -g
+-- corepack@0.10.0
 -- npm@8.3.1
```

`(.venv) >npm install -g ganache

Check again if Ganache was installed
```
(.venv) >npm list -g
+-- corepack@0.10.0
+-- ganache@7.0.1
 -- npm@8.3.1
```
 
Now Ganache should run in the virtual environment.
`(.venv) >ganache`

The whole flow without comments:
```
testEnv>python -m venv .venv
testEnv>.\.venv\Scripts\Activate.ps1
(.venv) testEnv>pip install nodeenv
  Collecting nodeenv
  Using cached nodeenv-1.6.0-py2.py3-none-any.whl (21 kB)
  Installing collected packages: nodeenv
  Successfully installed nodeenv-1.6.0
(.venv) testEnv>nodeenv -p
  * Install prebuilt node (17.4.0) ..... done.
  * Overwriting C:\testEnv\.venv\Scripts\activate.bat with new content
  * Overwriting C:\testEnv\.venv\Scripts\deactivate.bat with new content
  * Overwriting C:\testEnv\.venv\Scripts\Activate.ps1 with new content  
  C:\testEnv\.venv\Scripts\nodejs.exe <<===>> node.exe
(.venv) testEnv>npm root -g
  C:\testEnv\.venv\Scripts\node_modules
(.venv) testEnv>npm list -g
  C:\testEnv\.venv\Scripts
  +-- corepack@0.10.0
  `-- npm@8.3.1

(.venv) testEnv>npm install -g ganache

added 29 packages, and audited 30 packages in 51s

found 0 vulnerabilities
(.venv) testEnv>npm list -g
  C:\testEnv\.venv\Scripts
  +-- corepack@0.10.0
  +-- ganache@7.0.1
  `-- npm@8.3.1

(.venv) testEnv>ganache
ganache v7.0.1 (@ganache/cli: 0.1.2, @ganache/core: 0.1.2)
Starting RPC server

Available Accounts
==================
(0) 0xBED9BEEEA4ab4A8dD724660aDeBcAC187D139073 (1000 ETH)
```

**Comment by cromewar** [comment]
If you made an article (on medium or any other platform) we can tag this on the course index.
BTW you are doing just an amazing job helping out with a lot of cool tips @n4n0b1t3 :D

**Comment by n4n0b1t3** [comment]
Yes, I have a medium account. Would you suggest posting articles on medium instead? That might be better because I believe not many people are using the search function of the discussion board, judging by the repetitive nature of some posts.

**Reply by cromewar** [reply]
Yes , the forum is mostly used to make question about problems and some of this incredible post could get missed. So I strongly suggest create Medium articles of this, the information is really valuable.
